### ISSUE TITLE
Unable to install cert-manager for KF 1.3

### ISSUE THREAD

**Issue by nakfour** [kubeflow/manifests#1810]
while running the commands provided for cert-manager here : https://github.com/kubeflow/manifests#install-individual-components
I bumped into a couple of issues:
1. Both cert-manager and cert-manager-cainjector pods are trying to access resources in namespace kube-system. Looking at the deployment config for cert-manager, I see the ```leader-election-namespace=kube-system``` If Kubeflow installs cert-manager in namespace cert-manager, why is the leader-election-namespace kube-system. This is similar to this issue: https://github.com/jetstack/cert-manager/issues/1068
2. cert-manager-webhook pods fails on ```Unable to attach or mount volumes: unmounted volumes=[certs], unattached volumes=[certs cert-manager-webhook-token-qhb7b]: timed out waiting for the condition```

**Comment by DavidSpek** [comment]
@nakfour While this is a problem, the version of cert-manager in this repository is v0.11 while the latest is v1.2. I think you might be better off just installing cert-manager from the upstream manifests (I have a kustomize manifest for this [here](https://github.com/kubeflow-onprem/ArgoFlow/tree/master/cert-manager)).

@yanniszark I think this is the second issue related to cert-manager that I have seen here. I've brought this up before, but maybe we should update the cert-manager manifests in this repo. I can open a PR for this instantly seeing as I have the kustomize manifests for this already.

**Comment by nakfour** [comment]
@DavidSpek I am not sure an updated version of cert-manager will fix this, it might. The issue is with the namespace fields specified in this on the lines referring to kube-system namespace:  https://github.com/kubeflow/manifests/blob/master/common/cert-manager/cert-manager/base/deployment.yaml#L25
https://github.com/kubeflow/manifests/blob/master/common/cert-manager/cert-manager/base/deployment.yaml#L64

**Comment by yanniszark** [comment]
@nakfour thanks for the report! I would like to understand how the environment you are deploying inside is causing cert-manager to fail. First of all, could you confirm that you have all of these three kustomizations installed, as mentioned in the install guide (https://github.com/kubeflow/manifests/tree/v1.3.0-rc.0#cert-manager):

```yaml
# Cert-Manager
- ../common/cert-manager/cert-manager-kube-system-resources/base
- ../common/cert-manager/cert-manager-crds/base
- ../common/cert-manager/cert-manager/overlays/self-signed
```

I have already confirmed that installation to work inside a Kubernetes cluster, so something must be different in your environment and we need to find out what it is.

**Comment by nakfour** [comment]
@yanniszark yes my cluster is different, it is on Openshift that does not allow a service to query resources in another namespace and that is why we can see the issue of cert-manager installed on one namespace "cert-manager" and trying to use another namespace "kube-system". If the intention of KF is to install cert-manager in "cert-manager" namespace then this should be fixed. And yes all these Kustomizations are installed as specified in the install guide.

**Comment by DavidSpek** [comment]
@nakfour Out of curiosity, does `kustomize build github.com/kubeflow-onprem/ArgoFlow/cert-manager | kubectl apply -f -` work for you on OpenShift?

**Comment by nakfour** [comment]
@yanniszark @DavidSpek changing those two lines to use "cert-manager" namespace did fix the issue. I can include the change to cert-manager deployment in the Openshift PR #1811 if the community is okay with this?

**Comment by DavidSpek** [comment]
@nakfour I believe the manifests for cert-manager in this repository were just grabbed from cert-manager upstream in October of 2019. I don't think it is advisable to edit the cert-manager manifests directly as that is how cert-manager 0.11 is expecting them to be. I think the solution is to update cert-manager, as the new manifests are stated to work with OpenShift out of the box.

**Comment by yanniszark** [comment]
> @yanniszark yes my cluster is different, it is on Openshift that does not allow a service to query resources in another namespace and that is why we can see the issue of cert-manager installed on one namespace "cert-manager" and trying to use another namespace "kube-system". If the intention of KF is to install cert-manager in "cert-manager" namespace then this should be fixed. And yes all these Kustomizations are installed as specified in the install guide.

@nakfour thanks for the clarification. So you are saying that the cert-manager Pods can't create ConfigMaps in the kube-system namespace, even though there are RBAC rules allowing them to do so? In that case, I have a final question: Since cert-manager is the same as in previous Kubeflow releases, how did Openshift work in previous releases?

> I can include the change to cert-manager deployment in the Openshift PR #1811 if the community is okay with this?

I'm not sure how this worked, since the cert-manager Pods don't have RBAC permission to read / edit ConfigMaps in namespace `cert-manager`. Did you make any further changes?

**Comment by DavidSpek** [comment]
@yanniszark I think the solution is here obvious, and it is to update cert-manager. I also wonder why this was not done by the manifests working group in the past months as it only took me a maximum of 2 minutes to make the kustomize manifests for the latest version of cert-manager, and most of that time was waiting for the cert-manager website to load to get the deployment URL. I also brought up that cert-manager should be updated months ago, and not taking action then is causing this issue right before the release.

**Comment by nakfour** [comment]
@yanniszark looking at the rbacs: https://github.com/kubeflow/manifests/blob/v1.2-branch/cert-manager/cert-manager/base/cluster-role.yaml
It does not include configmap resource, and the pods where trying to query configmaps in namespace "kube-system" . The only thing we did differently for KF 1.2 is add the "overlays/application/application.yaml" in our Kustomize, not sure if this made any changes. Either way, it is good practice to point to the accurate namespace in the two lines I pointed above.

**Comment by yanniszark** [comment]
> @yanniszark I think the solution is here obvious, and it is to update cert-manager. I also wonder why this was not done by the manifests working group in the past months as it only took me a maximum of 2 minutes to make the kustomize manifests for the latest version of cert-manager, and most of that time was waiting for the cert-manager website to load to get the deployment URL. I also brought up that cert-manager should be updated months ago, and not taking action then is causing this issue right before the release.

@DavidSpek this solution is not obvious to me. When an issue comes up, we should try to understand its cause. This is the first step we should take regardless of the solution. As for upgrading cert-manager, I will repeat here that since we are at the end of a release, we are only including bug fixes. I agree that a new cert-manager is a good addition, but I would not make it a release blocker until we get to the bottom of the issue and understand what is going on. Please keep in mind that this is a version that has been extensively tested and working in previous versions. With that said, I am returning the discussion to understanding the root cause of this issue.

> It does not include configmap resource, and the pods where trying to query configmaps in namespace "kube-system" . The only thing we did differently for KF 1.2 is add the "overlays/application/application.yaml" in our Kustomize, not sure if this made any changes. Either way, it is good practice to point to the accurate namespace in the two lines I pointed above.

The RBAC resources for the kube-system namespace are defined in:
https://github.com/kubeflow/manifests/tree/master/common/cert-manager/cert-manager-kube-system-resources/base
Could you point me to the kustomizations you are using to install?

> @yanniszark another issue is with testing "platform-agnostic" for pipeline, was that tested? Its missing for example the "Viewer" CRD

We mainly tested with `platform-agnostic-multi-user`. This sounds like an issue for KFP, could you file an issue there so the KFP people can get to the bottom of it?

**Comment by DavidSpek** [comment]
@yanniszark This old version of cert-manager is probably extensively tested with older versions of Kubernetes, but not the latest releases. Furthermore, I find the choice to debug a severely outdated version of a third-party application quite interesting as issues with that application should be handled by the application developers, unless the manifests working group rather fork, fix and maintain an old version of cert-manager. I think the better choice would have been to remove the cert-manager manifests if the manifest working group did not actually want to maintain them, which seems to be the case since I brought up updating the cert-manager manifests months ago when there was more than enough time and this was rejected.

**Comment by nakfour** [comment]
@yanniszark these are roles and not cluster roles, so they are bound to a namespace, and hence why cert-manager should not be allowed to access configmap resources in "kube-system" namespace.  https://github.com/kubeflow/manifests/blob/master/common/cert-manager/cert-manager-kube-system-resources/base/role.yaml#L3
Here is our cert-manager, which is exactly as the installation instructions https://github.com/nakfour/manifests/blob/openshift-kf-13/distributions/stacks/openshift/application/cert-manager/kustomization.yaml
@yanniszark can you please answer my question, why cert-manager is trying to access namespace "kube-system" while it is installed in "cert-manager"

**Comment by yanniszark** [comment]
> Furthermore, I find the choice to debug a severely outdated version of a third-party application quite interesting as issues with that application should be handled by the application developers, unless the manifests working group rather fork, fix and maintain an old version of cert-manager.

I think you are misunderstanding what I am saying. I said that we should understand the reason why our current cert-manager manifests are failing on this particular instance, when they are succeeding everywhere else. After we find that out, we can discuss on the best way forward.

> I brought up updating the cert-manager manifests months ago when there was more than enough time and this was rejected.

I remember an issue about cert-manager failing in <1.16 installations, which turned out to be a misinterpretation of cert-manager's docs (looking at the wrong version). I don't remember an issue for upgrading (and I didn't find anything by searching). As I said, upgrading cert-manager is a good addition and something we want to do in the manifests wg in the immediate future, after the release. Since you are invested in this feature, would you like to open an issue to track it? It may be that we eventually decide that putting the effort to upgrade is the best way forward. But for now, let's keep this issue focused to locating the root cause. Please remember that no new features go in 1.3.0, only bugfixes at this point.

> @yanniszark can you please answer my question, why cert-manager is trying to access namespace "kube-system" while it is installed in "cert-manager"

I believe that cert-manager:
- Needs to create a ConfigMap to perform leader election for each component and creates them in kube-system.
- The cert-manager APIService needs to bind to the `extension-apiserver-authentication-reader` Role. See: https://kubernetes.io/docs/tasks/extend-kubernetes/setup-extension-api-server/

> Here is our cert-manager, which is exactly as the installation instructions https://github.com/nakfour/manifests/blob/openshift-kf-13/distributions/stacks/openshift/application/cert-manager/kustomization.yaml

@nakfour thanks for pointing me to the installation files. I think your issue is this line:
https://github.com/nakfour/manifests/blob/openshift-kf-13/distributions/stacks/openshift/application/cert-manager/kustomization.yaml#L7
Which is changing the namespace for the Role and RoleBinding of the `cert-manager-kube-system-resources` kustomization.

Can you please remove that line and tell me if the problem is gone?

**Comment by nakfour** [comment]
Thanks @yanniszark yes that made it work. However, I am still not clear that leader election has to happen in kube-system namespace. According to last comment in jetstack/cert-manager#1068 you can override the namespace where leader election happens. I just don't find it a good practice to install in one namespace and create all these resources for leader election in another. Either way we can close this issue for now and maybe move this discussion to @DavidSpek upgrade PR

**Comment by yanniszark** [comment]
> I just don't find it a good practice to install in one namespace and create all these resources for leader election in another.

We should definitely keep this data point as an argument for upgrading cert-manager.

> maybe move this discussion to @DavidSpek upgrade PR

Sure! I think it would be better to track discussions in issues, since we use issues to track work items.
@DavidSpek do you want to open an issue for cert-manager? If not, I can do it as well.

Thanks @nakfour! Closing this issue then, as Openshift is unblocked on this front.

**Comment by DavidSpek** [comment]
@yanniszark I can create another issue about cert-manager, but I would first like to coordinate merging https://github.com/kubeflow/kubeflow/pull/5823 quickly so we can close https://github.com/kubeflow/kubeflow/issues/5821 which is a bigger problem.
